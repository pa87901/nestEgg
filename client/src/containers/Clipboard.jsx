import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Segment, Button } from 'semantic-ui-react';
import fetch from 'isomorphic-fetch';
import Jotter from '../components/Jotter';
import Ticket from '../components/Ticket';
import {
  setTransactions,
  removeTransactions
} from '../actions/transactionActions';
import { updateHoldings } from '../actions/holdingActions';


class Clipboard extends Component {
  constructor() {
    super();
    this.delete = this.delete.bind(this);
  }

  componentWillMount() {
    const { handleSetTransactions } = this.props;
    fetch('http://localhost:3000/api/transactions', { method: 'get' })
    .then(transactions => transactions.json())
    .then(transactionsJSON => {
      handleSetTransactions(transactionsJSON);
    })
    .catch(err => {
      console.error('Unable to get all transactions from the backend:', err); // eslint-disable-line no-console
    });
  }

  delete() {
    const { selectedTransactions, handleRemoveTransactions, handleUpdateHoldings } = this.props;
    if (!selectedTransactions.length) {
      window.alert('No transactions selected. Please select transactions by checkboxes.');
      return;
    }
    console.log('selectedTransactions:', selectedTransactions);
    const payload = { selectedTransactions };
    const headers = {
      'accept': 'application/json, text/plain, */*',
      'content-type': 'application/json'
    };
    const init = {
      method: 'DELETE',
      headers,
      mode: 'cors',
      body: JSON.stringify(payload),
      json: true
    };
    fetch('/api/transactions/', init)
    .then(res => res.json())
    .then(resJSON => {
      console.log('resJSON:', resJSON);
      const { responseSet } = resJSON;
      handleRemoveTransactions(selectedTransactions);
      handleUpdateHoldings(responseSet)
    })
    .catch(err => {
      console.error('Error sending ids to delete:', err) // eslint-disable-line no-console
    })
  }

  render() {
    const { selectedTransactions } = this.props;
    return (
      <Segment.Group>
        <Segment>
          <h2>Transactions</h2>
          <Jotter />
        </Segment>
        <Segment className="portfolioButtonRow">
          <Ticket />
          <Button
            basic
            color={selectedTransactions.length ? "red" : null}
            onClick={this.delete}>
            Delete
          </Button>
        </Segment>
      </Segment.Group>
    );
  }
}

Clipboard.propTypes = {
  handleSetTransactions: PropTypes.func.isRequired,
  selectedTransactions: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
  handleRemoveTransactions: PropTypes.func.isRequired,
  handleUpdateHoldings: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  selectedTransactions: state.transactions.selectedTransactions
});

const mapDispatchToProps = dispatch => ({
  handleSetTransactions(transactions) {
    dispatch(setTransactions(transactions));
  },
  handleRemoveTransactions(transactions) {
    dispatch(removeTransactions(transactions))
  },
  handleUpdateHoldings(holdings) {
    dispatch(updateHoldings(holdings));
  }
});

export const Unwrapped = Clipboard;
export default connect(mapStateToProps, mapDispatchToProps)(Clipboard);