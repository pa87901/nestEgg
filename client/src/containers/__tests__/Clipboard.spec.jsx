import React from 'react';
import { shallow } from 'enzyme';
import { Unwrapped as UnwrappedClipboard } from '../Clipboard';

describe('Clipboard', () => {
  const shallowComponent = shallow(<UnwrappedClipboard store={[]} handleSetTransactions={() => {}} />);

  test('Clipboard snapshot has not changed', () => {
    expect(shallowComponent).toMatchSnapshot();
  });
});