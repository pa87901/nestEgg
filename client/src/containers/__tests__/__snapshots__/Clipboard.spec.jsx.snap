// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clipboard Clipboard snapshot has not changed 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Clipboard
    handleSetTransactions={[Function]}
    store={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Segment>
          <h2>
            Transactions
          </h2>
          <Connect(Jotter) />
        </Segment>,
        <Segment
          className="portfolioButtonRow"
        >
          <Connect(Ticket) />
          <Button
            as="button"
            basic={true}
            color={null}
          >
            Delete
          </Button>
        </Segment>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <h2>
              Transactions
            </h2>,
            <Connect(Jotter) />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Transactions",
            },
            "ref": null,
            "rendered": "Transactions",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Connect(Ticket) />,
            <Button
              as="button"
              basic={true}
              color={null}
            >
              Delete
            </Button>,
          ],
          "className": "portfolioButtonRow",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "basic": true,
              "children": "Delete",
              "color": null,
            },
            "ref": null,
            "rendered": "Delete",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Segment>
            <h2>
              Transactions
            </h2>
            <Connect(Jotter) />
          </Segment>,
          <Segment
            className="portfolioButtonRow"
          >
            <Connect(Ticket) />
            <Button
              as="button"
              basic={true}
              color={null}
            >
              Delete
            </Button>
          </Segment>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <h2>
                Transactions
              </h2>,
              <Connect(Jotter) />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Transactions",
              },
              "ref": null,
              "rendered": "Transactions",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Connect(Ticket) />,
              <Button
                as="button"
                basic={true}
                color={null}
              >
                Delete
              </Button>,
            ],
            "className": "portfolioButtonRow",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "basic": true,
                "children": "Delete",
                "color": null,
              },
              "ref": null,
              "rendered": "Delete",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
